list unmounted mbeds on Linux.  closes #56

justbuchanan opened the pull request.

bridadan commented:
This would probably break some existing tools that rely on being able to flash the device if it shows up in mbed-ls, correct?

Maybe this behavior should be turned on with a flag or something? I do think this is still a useful feature, just concerned about compatibility

justbuchanan commented:
That's a fair point, I guess some users could be relying on anything that's returned being already mounted.

I just pushed a commit to this branch so that unmounted mbeds are only listed if the "-u" flag is passed.  The OS X and Windows implementations essentially ignore the flag since on those platforms the MBED will always automatically mount (as far as I know).

bridadan commented:
Looks good to me!

justbuchanan commented:
Do you know why circleci is failing?  This is the error message, it doesn't seem to be because of my changes:

~~~{.sh}
INFO:coveralls:{"message":"Couldn't find a repository matching this job.","error":true} coveralls returned exit code 1
~~~

bridadan commented:
Can't say I do, seems to be an infrastructure problem? cc @PrzemekWirkus

PrzemekWirkus commented:
It fails because connection to [```coveralls``` service](https://coveralls.io/github/ARMmbed/mbed-ls?branch=master) died for this build. It hap[pens from time to time, I need to investigate why.

PrzemekWirkus commented:
BTW, this feature is not so bad ;)
* And it is optional with ```-u``` switch so doesn't break existing setup.
* I would definitely move this flag from method parameter list to class as class member.
* It would be nice to have some kind of warning that there is a mbed device connected but it is not mounted! :)

justbuchanan commented:
Thanks!  I moved the parameter to be an instance variable as suggested and I think it's much cleaner that way.  I also added a warning for when an mbed is omitted due to being disconnected.  Let me know if there's anything else I should change.

justbuchanan commented:
@PrzemekWirkus is there anything else I need to do for this pull request?

PrzemekWirkus commented:
@justbuchanan Please be patient. I'm quite busy now so I need to push mbedls down the list.

PrzemekWirkus commented:
@justbuchanan Actually you could add to README (for example under this chapter: https://github.com/ARMmbed/mbed-ls#mbedls-as-command-line-tool) description with example for ```-u``` switch.
For example you can show before and after mount mbedls output. And explain why we've added it.

justbuchanan commented:
@PrzemekWirkus no problem, I understand.   I added a short description to the readme, let me know if this suffices.

PrzemekWirkus commented:
@justbuchanan 
I was checking this PR before merge and I've noticed that there are some problems.
So I've plugged two mbed devices:

```
przemek@ubuntu:~/work/justbuchanan/mbed-ls$ mbedls
+---------------+----------------------+-------------------------+--------------+--------------------------------------------------+-----------------+
| platform_name | platform_name_unique | mount_point             | serial_port  | target_id                                        | daplink_version |
+---------------+----------------------+-------------------------+--------------+--------------------------------------------------+-----------------+
| K64F          | K64F[0]              | /media/przemek/DAPLINK1 | /dev/ttyACM1 | 0240000033514e45001f500585d40014e981000097969900 | 0241            |
| NUCLEO_F401RE | NUCLEO_F401RE[0]     | /media/przemek/NUCLEO   | /dev/ttyACM0 | 07200200076165023804F31F                         | 0200            |
+---------------+----------------------+-------------------------+--------------+--------------------------------------------------+-----------------+
```

and umounted one of them (NUCLEO):
```
przemek@ubuntu:~/work/justbuchanan/mbed-ls$ mbedls -u
+---------------+----------------------+-------------------------+--------------+--------------------------------------------------+-----------------+
| platform_name | platform_name_unique | mount_point             | serial_port  | target_id                                        | daplink_version |
+---------------+----------------------+-------------------------+--------------+--------------------------------------------------+-----------------+
| K64F          | K64F[0]              | /media/przemek/DAPLINK1 | /dev/ttyACM1 | 0240000033514e45001f500585d40014e981000097969900 | 0241            |
+---------------+----------------------+-------------------------+--------------+--------------------------------------------------+-----------------+
przemek@ubuntu:~/work/justbuchanan/mbed-ls$ ll /dev/disk/by-id/
total 0
drwxr-xr-x 2 root root 120 Apr  8 05:08 ./
drwxr-xr-x 6 root root 120 Mar 31 02:43 ../
lrwxrwxrwx 1 root root   9 Mar 31 02:43 ata-VMware_Virtual_SATA_CDRW_Drive_00000000000000000001 -> ../../sr0
lrwxrwxrwx 1 root root   9 Mar 31 02:43 ata-VMware_Virtual_SATA_CDRW_Drive_01000000000000000001 -> ../../sr1
lrwxrwxrwx 1 root root   9 Apr  8 05:08 usb-MBED_microcontroller_0672FF485649785087171742-0:0 -> ../../sdb
lrwxrwxrwx 1 root root   9 Apr  8 05:07 usb-MBED_VFS_0240000033514e45001f500585d40014e981000097969900-0:0 -> ../../sdc
przemek@ubuntu:~/work/justbuchanan/mbed-ls$ ll /dev/serial/by-id/
total 0
drwxr-xr-x 2 root root 80 Apr  8 05:08 ./
drwxr-xr-x 4 root root 80 Apr  8 05:07 ../
lrwxrwxrwx 1 root root 13 Apr  8 05:07 pci-ARM_DAPLink_CMSIS-DAP_0240000033514e45001f500585d40014e981000097969900-if01 -> ../../ttyACM1
lrwxrwxrwx 1 root root 13 Apr  8 05:08 pci-STMicroelectronics_STM32_STLink_0672FF485649785087171742-if02 -> ../../ttyACM0
przemek@ubuntu:~/work/justbuchanan/mbed-ls$ mount | grep vfat
/dev/sdc on /media/przemek/DAPLINK1 type vfat (rw,nosuid,nodev,relatime,uid=1000,gid=1000,fmask=0022,dmask=0022,codepage=437,iocharset=iso8859-1,shortname=mixed,showexec,utf8,flush,errors=remount-ro,uhelper=udisks2)
```

Debug info after umount:
```
przemek@ubuntu:~/work/justbuchanan/mbed-ls$ mbedls -d
debug @MbedLsToolsUbuntu.get_dev_by_id_cmd: ls -oA /dev/disk/by-id/
debug @MbedLsToolsUbuntu.get_dev_by_id_process: lrwxrwxrwx 1 root 9 Mar 31 02:43 ata-VMware_Virtual_SATA_CDRW_Drive_00000000000000000001 -> ../../sr0
debug @MbedLsToolsUbuntu.get_dev_by_id_process: lrwxrwxrwx 1 root 9 Mar 31 02:43 ata-VMware_Virtual_SATA_CDRW_Drive_01000000000000000001 -> ../../sr1
debug @MbedLsToolsUbuntu.get_dev_by_id_process: lrwxrwxrwx 1 root 9 Apr  8 05:08 usb-MBED_microcontroller_0672FF485649785087171742-0:0 -> ../../sdb
debug @MbedLsToolsUbuntu.get_dev_by_id_process: lrwxrwxrwx 1 root 9 Apr  8 05:07 usb-MBED_VFS_0240000033514e45001f500585d40014e981000097969900-0:0 -> ../../sdc
debug @MbedLsToolsUbuntu.get_dev_by_id_cmd: ls -oA /dev/serial/by-id/
debug @MbedLsToolsUbuntu.get_dev_by_id_process: lrwxrwxrwx 1 root 13 Apr  8 05:07 pci-ARM_DAPLink_CMSIS-DAP_0240000033514e45001f500585d40014e981000097969900-if01 -> ../../ttyACM1
debug @MbedLsToolsUbuntu.get_dev_by_id_process: lrwxrwxrwx 1 root 13 Apr  8 05:08 pci-STMicroelectronics_STM32_STLink_0672FF485649785087171742-if02 -> ../../ttyACM0
debug @MbedLsToolsUbuntu.get_mounts: mount | grep vfat
debug @MbedLsToolsUbuntu.get_mounts: /dev/sdc on /media/przemek/DAPLINK1 type vfat (rw,nosuid,nodev,relatime,uid=1000,gid=1000,fmask=0022,dmask=0022,codepage=437,iocharset=iso8859-1,shortname=mixed,showexec,utf8,flush,errors=remount-ro,uhelper=udisks2)
debug @MbedLsToolsUbuntu.scan_html_line_for_target_id: window.location.replace("https://mbed.org/device/?code=0240000033514e45001f500585d40014e981000097969900?version=0241?target_id=001bffffffffffff4e45315400050003");
debug @MbedLsToolsUbuntu.scan_html_line_for_target_id: (('0240000033514e45001f500585d40014e981000097969900',), '0240000033514e45001f500585d40014e981000097969900')
debug @MbedLsToolsUbuntu.list_mbeds_ext: ('K64F[0]', '0240000033514e45001f500585d40014e981000097969900')
+---------------+----------------------+-------------------------+--------------+--------------------------------------------------+-----------------+
| platform_name | platform_name_unique | mount_point             | serial_port  | target_id                                        | daplink_version |
+---------------+----------------------+-------------------------+--------------+--------------------------------------------------+-----------------+
| K64F          | K64F[0]              | /media/przemek/DAPLINK1 | /dev/ttyACM1 | 0240000033514e45001f500585d40014e981000097969900 | 0241            |
+---------------+----------------------+-------------------------+--------------+--------------------------------------------------+-----------------+
debug @MbedLsToolsUbuntu.mbed_lstools.main: Return code: 0
```

justbuchanan commented:
I tried reproducing this on my computer and I couldn't get it to break.  I used two LPC1768's though since that's all I have access to.  Could it be something specific to the NUCLEO?  If you get a chance, could you try this again using a device other than the NUCLEO and see what the results are?

It's also possible it's something ubuntu-related since it works on my computer and I'm running a different distro.  I'll load up an ubuntu virtual machine and see what happens.

justbuchanan commented:
@PrzemekWirkus have you had a chance to look at this again?  I still can't reproduce the issue you were having :/

jupe commented:
This could help recover some scenarios in CI, so please continue working with this :)

mazimkhan commented:
@justbuchanan Can you please rebase with master.

justbuchanan commented:
@mazimkhan done

adbridge commented:
@mazimkhan are you happy with this now ?

bridadan commented on mbed_lstools/main.py:
```diff
@@ -170,6 +176,8 @@ def get_mbedls_version():
     mbeds.debug(__name__, "mbed-ls ver. " + get_mbedls_version())
     mbeds.debug(__name__, "host: " +  str((mbed_lstools_os_info())))
 
+    mbeds.list_unmounted = opts.list_unmounted
```
This feels strange, no other option in the CLI modifies the `mbeds` object directly.

Instead, can this option be passed into the `list` functions directly as an argument?

theotherjimmy replied:
Could you remove this assignment?

theotherjimmy commented on mbed_lstools/lstools_linux_generic.py:
```diff
@@ -111,7 +113,12 @@ def list_mbeds(self):
             mbed['target_id_usb_id'] = tid
             mbed['target_id_mbed_htm'] = mbed_htm_target_id
-            result.append(mbed)
+            if mbed['mount_point'] != None or self.list_unmounted:
```
Could you change this check to:
```python
if mbed['mount_point'] or self.list_unmounted:
```

I think that's more idiomatic python. It should also catch some odd corner cases like the empty string as a mount point.

theotherjimmy commented:
@justbuchanan Do you have time/want to update this patch? Would you like me to take it over?

justbuchanan commented:
@theotherjimmy feel free to take over this PR

theotherjimmy commented:
Will do. Thanks for the initial implementation. I'm going to close this PR, and move the branch to my fork.
